### PR TITLE
[Feature] Support serialization for smart pointer

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
What is the use case of this feature?

**Comment by VoVAllen** [comment]
Previous implementation includes bugs and the smart pointer needs to be handled separately. For example `vector<shared_ptr<T>>` can not directly be serialized as `vector<T>` could. Here we want to make both of them works.

File: include/dgl/runtime/smart_ptr_serializer.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+template <typename T>
+struct Handler<std::shared_ptr<T>> {
+  inline static void Write(Stream *strm, const std::shared_ptr<T> &data) {
+    Handler<T>::Write(strm, *data.get());
+  }
+  inline static bool Read(Stream *strm, std::shared_ptr<T> *data) {
+    // When read, the default initialization behavior of shared_ptr is
+    // shared_ptr<T>(), which is holding a nullptr. Here we need to manually
+    // reset to a real object for further loading
+    if (data->get() == nullptr) {
```
Comment (jermainewang) [inline include/dgl/runtime/smart_ptr_serializer.h]: `if (!data)`

File: include/dgl/runtime/smart_ptr_serializer.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+template <typename T>
+struct Handler<std::unique_ptr<T>> {
+  inline static void Write(Stream *strm, const std::unique_ptr<T> &data) {
+    Handler<T>::Write(strm, *data.get());
+  }
+  inline static bool Read(Stream *strm, std::unique_ptr<T> *data) {
+    // When read, the default initialization behavior of unique_ptr is
+    // unique_ptr<T>(), which is holding a nullptr. Here we need to manually
+    // reset to a real object for further loading
+    if (data->get() == nullptr) {
```
Comment (jermainewang) [inline include/dgl/runtime/smart_ptr_serializer.h]: ditto

**Pull request merged**